### PR TITLE
Add birthday field to contact

### LINKED ISSUE #5: No birthday field
There's no field to add someone's birthday, there should be this option, and it should sync with the calendar to show someone's birthday is on a certain day.

### PR DESCRIPTION
**Pull request by lduboeuf**
Add a birthday field in contact. this fixes https://github.com/ubports/address-book-app/issues/5

The feature is already testable . Birthday are synced with calendar as expected, but note that there is a known bug waiting to be merged  , the date set and the date displayed after saving might be different depending of the timezone : https://github.com/ubports/address-book-service/pull/12

~~Also, another issue that doesn't make calendar pointing to the right day is also not merged yet: see : https://gitlab.com/ubports/apps/calendar-app/-/merge_requests/187~~

testable click here: https://drive.google.com/open?id=1JVPkuDY10Qh7fEYkUAIskuR2c9K7b2ah
( recommended way to test because of others related PRs included )

### CODE REVIEW

**Comment by Danfro** [comment]
This is so great! Tested on E5 RC with install-qa.

Three issues I do encounter:
1. edit contact and save a birthday, say today 14.05.2020, when checking the date it is out by one day, in my example the 13.05. will be stored (and opened in calendar). I guess that is due to the timezone issue mentioned above [ 0:00 -> 23:00].

2. Delete an existing birthday, you can not immediately add a birthday again because the "birthday" field is not available. You need to close the edit session and reopen the contact. 

3. An birthday can only be edited once. Save birthday, close edit session. Reopen edit session, change birthday value and press the "tick" icon. The UI will freeze.
Error in log:
ContactEditorPage.qml:190:TypeError: Cannot read property "onContactChanged" of null
Great job.

**Comment by lduboeuf** [comment]
thank you,
Well the 1st issue is fixed by another PR waiting to be merged
The 2nd is a general issue, not related to birthday addition, i will look at it
Will see first for the 3rd one, but i don't reproduce it. Is it a special Contact, imported from elsewhere, what kind of fields are set ?

**Comment by Danfro** [comment]
Hm. I rebooted, just to be sure.

Now I
1. can not reproduce the number 3 issue mentioned above anymore
2. can confirm that the date issue is fixed by PR [#12](https://github.com/ubports/address-book-service/pull/12) 

I did install both PR's. I am not 100% sure. I might not have rebooted after installing the second one. Maybe that is why. So all fine now.

File: po/address-book-app.pot [hunk 1]
```diff
@@ -148,6 +154,10 @@ msgstr ""
 msgid "Email"
 msgstr ""
 
+#: ../src/imports/Ubuntu/AddressBook/ContactEditor/ContactDetailBirthdayEditor.qml:129
+msgid "Enter a Birthday"
```
Comment (Flohack74) [inline po/address-book-app.pot]: birthday should be lowercase here

File: src/imports/Ubuntu/AddressBook/ContactEditor/ContactEditorPage.qml [hunk 2]
```diff
@@ -169,14 +170,19 @@ Page {
...
     function focusToFirstEntry(field)
     {
-        var itemToFocus = field
-        if (field.repeater)
-            itemToFocus = field.repeater.itemAt(0)
-
-        if (itemToFocus) {
-            root.idleMakeMeVisible(itemToFocus)
-            itemToFocus.forceActiveFocus()
+        //only focus when new field requested
```
Comment (Flohack74) [inline src/imports/Ubuntu/AddressBook/ContactEditor/ContactEditorPage.qml]: Is this an intentional change? Then you should also describe it in your PR. Otherwise this is a kind of "drive-by" change that we want to prevent normally. In some cases you would even need to split it off into another PR. So make your PR text "on-topic" with that change :)

**Review by Flohack74 — CHANGES_REQUESTED** [review]
See comments

**Comment by lduboeuf** [comment]
> Delete an existing birthday, you can not immediately add a birthday again because the "birthday" field is not available. You need to close the edit session and reopen the contact.

Humm that simple UI thing is quite complex to handle as the option list is populated according to the model. When editing a contact, nothing is done to the model until we validate the form.
This would lead to a complete rewrite of the editing management. E.g removing a field would effectively remove something in the model or a copy of it.
 A simple way would be to always have the "birthday" option always available even if there is already one.
What do you think ?
[EDIT] : I've added that feature, if you wish to test it

File: src/imports/Ubuntu/AddressBook/ContactEditor/ContactEditorPage.qml (hunk 2 above)
Comment (lduboeuf) [reply src/imports/Ubuntu/AddressBook/ContactEditor/ContactEditorPage.qml]: Humm, in fact, it was not a visual issue since adding that special birthday field even if all the forceFocus things were unnecessarily fired . You're right, I will propose another PR for that and merge it into the click test to avoid feedbacks about it :)

Comment (lduboeuf) [reply src/imports/Ubuntu/AddressBook/ContactEditor/ContactEditorPage.qml]: can you please merge the clickable https://github.com/ubports/address-book-app/pull/121 so that i have to avoid always cherry picking "-n" from it ?

**Comment by bhdouglass** [comment]
@Flohack74 and/ore @lduboeuf what's the status here?

**Comment by lduboeuf** [comment]
it is ready on my side, waiting to be merged. But need to have feedbacks with click tests

**Comment by lduboeuf** [comment]
~~humm i see issues with current test click ( i had the nb 3 of https://github.com/ubports/address-book-app/pull/128#issuecomment-628841120 but random) 
Will have a look at it~~

Note that PR in address-book-service is merged but not published yet, so we can have issues with Date Timezone

File: src/imports/Ubuntu/AddressBook/ContactView/ContactDetailBirthdayView.qml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2012-2015 Canonical, Ltd.
```
Comment (Flohack74) [inline src/imports/Ubuntu/AddressBook/ContactView/ContactDetailBirthdayView.qml]: Is this class really coming from Canonical? If not you can set (c) 2020 UBports Foundation ^^ or your name

Comment (lduboeuf) [reply src/imports/Ubuntu/AddressBook/ContactView/ContactDetailBirthdayView.qml]: yes true, lets go for Ubports foundation :)

**Review by Flohack74 — APPROVED** [review]

**Comment by Danfro** [comment]
We really should make sure we all use the same copyright's. I already have been adding some new files to weather app and used only "UBports". Example [here](https://gitlab.com/ubports/apps/weather-app/-/blob/master/app/ui/settings/TemperatureColorPage.qml#L2). I so far only used this for brand new files that I have added.

We need to make sure everyone merging MR's to core apps knows what to look for and gives appropriate instructions to contributors. @Flohack74 could this be discussed on the board? Then make a guidline and send it to all maintainers of core apps.

**Comment by Danfro** [comment]
And I would not mind changing my copyrights if needed. As long as we all have the same...

**Comment by Flohack74** [comment]
Well first of all the board is a little high in the hierarchy, and actually it would only be engaged if there is dispute over copyright ownership. And here comes the next problem, contributors normally retain their copyright and we cannot force anyone to give this up, only beg for it. So it would be perfectly normal that you put your name as copyright holder there.
How Qt and network-manager and also Canonical solved this is by enforcing contributors to sign CLA agreements, so that basically the copyright owner acknowledges that his work can be used in any way, he gives a perpetual license so to say and he cannot revoke this license. That would need us to have a process before someone contributes that he signs such a paper. Oh god, kinda too much for what we are doing.

So there is no real value in enforcing anything here, as we can only say ok please would you be so kind? But yes, we can bring this to the Developer Steering Committe. I am afraid however that we all will forget about this as its really a soft criteria and has not much consequences in the end :)

**Comment by Danfro** [comment]
Hm. You are right. Nothing that can be enforced. I am merely wishing for consistence. :smiley: I somehow get the feeling copyright is not doing me this favour. :sight:

So which string do we wish to use IF we do use a UBports copyright IF someone asks what to put in like here ^^? We should at least try to all tell the same story. UBports or UBports foundation or whatever the contributor likes/preferes?

**Comment by ghost** [comment]
Works fine one FP2. Thank you !  And the issues mentioned [above](https://github.com/ubports/address-book-app/pull/128#issuecomment-628841120) do not appear :)

**Comment by farkasdvd** [comment]
Device: bq e5
Image: 2020-W35/2

The birthday shows up correctly in the calendar, however it is not possible to edit the event there. Therefore it is not possible to set a reminder.
On the address-book-app side the birthday is not visible in dark mode. It shows only a white rectangle.

**Comment by lduboeuf** [comment]
For the not editable event. It is more an issue in calendar-app.
Will have a look for the dark mode issue

**Comment by farkasdvd** [comment]
@lduboeuf thanks, created an issue for calendar-app:
https://gitlab.com/ubports/apps/calendar-app/-/issues/194

**Comment by jezek** [comment]
Device: FP2
Image: 2020-W35/2

There is a birthday field in contact apps in contacts with filled birthday (in darkmode unreadable).

I use google contacts to sync, and some contacts have filled their birthday. Some contacts have filled full birhday (dd-mm-yyyy), some only day in year (dd-mm). The birthdays without year are not shown.

After tapping on birthday, the calendar opens on that day on tis year. I noticed, that there is reminder in calendar only for next birthday. The previous and the birthdays after next ar not listed in calendar (in google calendar they are).
The reminder in calendar is not editable, which is consistent with google calendar, cause you edit the birthday in contacts. 

When editing birthday in contacts app, in dark mode the day chooser/rotator has odd colors. After synchronizing, the date was changed correctly in calendar (after sync). Adding and removing birthday works too. Here I've spotted an oddity, sometimes after editing or deleting, the changes were not synchronized with google. Sometimes it worked, sometime not. But I think, it is not fault of this PR, it is some syncing bug.

**Comment by lduboeuf** [comment]
yes i've fixed the dark mode issue, hopefully going to be merged for OTA-13
Mind you give us screenshots for odd color ?

**Comment by lduboeuf** [comment]
yes i've fixed the dark mode issue, hopefully going to be merged for OTA-13
https://github.com/ubports/address-book-app/pull/153
Mind you give us screenshots for odd color ?

**Comment by jezek** [comment]
> Mind you give us screenshots for odd color ?

Of course, here it is.
screenshot20200901_221350019 resized 25%

**Comment by lduboeuf** [comment]
humm well i think there is a bug in Suru theme for Rectangle Qml Item
cc @cibersheep

**Comment by cibersheep** [comment]
> humm well i think there is a bug in Suru theme for Rectangle Qml Item
> cc @cibersheep

Are you using a default Picker?

**Comment by lduboeuf** [comment]
yes,
Ubuntu.Components.Pickers 1.3
 https://github.com/ubports/address-book-app/blob/xenial/src/imports/Ubuntu/AddressBook/ContactEditor/ContactDetailBirthdayEditor.qml#L86

**Comment by cibersheep** [comment]
Thank you. I'll take a look at it
For now, a workaround could to set StyleHint like Danfro did for the Calendar?:
https://gitlab.com/ubports/apps/calendar-app/-/merge_requests/175/diffs

**Comment by lduboeuf** [comment]
yes we should try with that thx, anyway this won't get into OTA-13 as i understand

**Comment by Danfro** [comment]
@cibersheep many thanks for digging out my MR for calendar app. Wanted to suggest that myself but did not find the time for it. :cry:

**Comment by Flohack74** [comment]
Funny I just checked a contact, it has already birthday inside, was that synced before already?

**Comment by lduboeuf** [comment]
> Funny I just checked a contact, it has already birthday inside, was that synced before already?

Yes, luckily the birthday sync was already in place before this PR

**Comment by Danfro** [comment]
But sadly it was using not the birthday, but the reminder day for the entry. I had a reminder set to 5 days before the birthday. That date got set as birthday. Resulted in some early congratulations until I found out the source of that... 😝

--

**Comment by lduboeuf** [comment]
> But sadly it was using not the birthday, but the reminder day for the entry. I had a reminder set to 5 days before the birthday. That date got set as birthday. Resulted in some early congratulations until I found out the source of that... stuck_out_tongue_closed_eyes
> […](#)
> --

Oh ok, that's another issue out of this scope i would say.

**Comment by Danfro** [comment]
True. Not adressed here. And I am not sure who to blame. My old vcf or dekko.

--

**Comment by jezek** [comment]
> But sadly it was using not the birthday, but the reminder day for the entry.

@Danfro Maybe you just answered a question that bugged me veeery long time. Why some birthdays are suddenly off by one day? This could be the answer. My remainders are set 1 day before a birthday and only for some contacts. Thanks.
Note: using google contacts to sync
